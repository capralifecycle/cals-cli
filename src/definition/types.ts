import { Permission } from '../github/types'

export interface Definition {
  snyk: {
    accountId: string
  }
  github: {
    users: User[]
    teams: {
      organization: string
      teams: Team[]
    }[]
  }
  projects: Project[]
}

export interface Project {
  name: string
  github: {
    organization: string
    repos?: DefinitionRepo[]
    teams?: RepoTeam[]
  }[]
}

export type User = UserBot | UserEmployee | UserExternal

export interface UserBot {
  type: 'bot'
  login: string
  name: string
}

export interface UserEmployee {
  type: 'employee'
  login: string
  capraUsername: string
  name: string
}

export interface UserExternal {
  type: 'external'
  login: string
  name: string
}

export interface DefinitionRepo {
  name: string
  archived?: boolean
  issues?: boolean
  wiki?: boolean
  teams?: RepoTeam[]
  snyk?: boolean
  public?: boolean
}

export interface RepoTeam {
  name: string
  permission: Permission
}

export interface Team {
  name: string
  members: string[] // Set
}
