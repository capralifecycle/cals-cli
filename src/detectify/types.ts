// See https://developer.detectify.com/#scan-profiles-manage-scan-profiles-get
export interface DetectifyScanProfile {
  name: string
  endpoint: string
  status: 'verified' | 'unverified' | 'unable_to_resolve' | 'unable_to_complete'
  created: string
  token: string
}

// See https://developer.detectify.com/#scan-reports-get-latest-report-get
export interface DetectifyScanReport {
  token: string
  scan_profile_token: string
  scan_profile_name: string
  created: string
  started: string
  stopped: string
  url: string
  cvss: number
  high_level_findings: number
  medium_level_findings: number
  low_level_findings: number
  information_findings: number
}
