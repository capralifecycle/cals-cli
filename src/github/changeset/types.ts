import { Permission } from "../types"

// GitHub repos

interface RepoCreateItem {
  type: "repo-create"
  org: string
  repo: string
}

interface RepoDeleteItem {
  type: "repo-delete"
  org: string
  repo: string
}

export interface RepoAttribUpdateItem {
  type: "repo-update"
  org: string
  repo: string
  attribs: (
    | { archived: boolean }
    | { issues: boolean }
    | { wiki: boolean }
    | { private: boolean }
  )[]
}

interface RepoTeamAddItem {
  type: "repo-team-add"
  org: string
  repo: string
  team: string
  permission: Permission
}

interface RepoTeamRemoveItem {
  type: "repo-team-remove"
  org: string
  repo: string
  team: string
}

interface RepoTeamPermissionItem {
  type: "repo-team-permission"
  org: string
  repo: string
  team: string
  permission: Permission
  current: {
    permission: Permission
  }
}

// GitHub members

interface MemberRemoveItem {
  type: "member-remove"
  org: string
  user: string
}

interface MemberAddItem {
  type: "member-add"
  org: string
  user: string
}

// GitHub teams

interface TeamRemoveItem {
  type: "team-remove"
  org: string
  team: string
}

interface TeamAddItem {
  type: "team-add"
  org: string
  team: string
}

interface TeamMemberRemoveItem {
  type: "team-member-remove"
  org: string
  team: string
  user: string
}

interface TeamMemberAddItem {
  type: "team-member-add"
  org: string
  team: string
  user: string
  role: "member" | "maintainer"
}

interface TeamMemberPermissionItem {
  type: "team-member-permission"
  org: string
  team: string
  user: string
  role: "member" | "maintainer"
}

/**
 * A change set item describes a transition on the end
 * service to become in sync with the desired definition.
 */
export type ChangeSetItem =
  | RepoCreateItem
  | RepoDeleteItem
  | RepoAttribUpdateItem
  | RepoTeamPermissionItem
  | RepoTeamAddItem
  | RepoTeamRemoveItem
  | MemberAddItem
  | MemberRemoveItem
  | TeamRemoveItem
  | TeamAddItem
  | TeamMemberRemoveItem
  | TeamMemberAddItem
  | TeamMemberPermissionItem
