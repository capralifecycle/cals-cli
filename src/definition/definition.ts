import fs from "fs"
import yaml from "js-yaml"
import { uniq } from "lodash"
import { Definition } from "./types"

function getTeamId(org: string, teamName: string) {
  return `${org}/${teamName}`
}

export function getRepoId(orgName: string, repoName: string) {
  return `${orgName}/${repoName}`
}

function validateDefinition(definition: Definition) {
  // Verify no duplicates in users and extract known logins.
  const loginList = definition.github.users.reduce<string[]>((acc, user) => {
    if (acc.includes(user.login)) {
      throw new Error(`Duplicate login: ${user.login}`)
    }
    return [...acc, user.login]
  }, [])

  // Verify no duplicates in teams and extract team names.
  const teamIdList = definition.github.teams.reduce<string[]>(
    (acc, orgTeams) => {
      return orgTeams.teams.reduce<string[]>((acc1, team) => {
        const id = getTeamId(orgTeams.organization, team.name)
        if (acc1.includes(id)) {
          throw new Error(`Duplicate team: ${id}`)
        }
        return [...acc1, id]
      }, acc)
    },
    [],
  )

  // Verify team members exists as users.
  definition.github.teams
    .map((it) => it.teams)
    .flat()
    .forEach((team) => {
      team.members.forEach((login) => {
        if (!loginList.includes(login)) {
          throw new Error(
            `Team member ${login} in team ${team.name} is not registered in user list`,
          )
        }
      })
    })

  // Verify no duplicates in project names.
  definition.projects.reduce<string[]>((acc, project) => {
    if (acc.includes(project.name)) {
      throw new Error(`Duplicate project: ${project.name}`)
    }
    return [...acc, project.name]
  }, [])

  definition.projects.forEach((project) => {
    project.github.forEach((org) => {
      // Verify project teams exists as teams.
      ;(org.teams || []).forEach((team) => {
        const id = getTeamId(org.organization, team.name)
        if (!teamIdList.includes(id)) {
          throw new Error(
            `Project team ${id} in project ${project.name} is not registered in team list`,
          )
        }
      })

      // Verify repo teams exists as teams.
      ;(org.repos || []).forEach((repo) => {
        ;(repo.teams || []).forEach((team) => {
          const id = getTeamId(org.organization, team.name)
          if (!teamIdList.includes(id)) {
            throw new Error(
              `Repo team ${id} for repo ${repo.name} in project ${project.name} is not registered in team list`,
            )
          }
        })
      })
    })
  })

  // Verify no duplicates in repos.
  definition.projects
    .flatMap((project) =>
      project.github
        .map((org) =>
          (org.repos || []).map((repo) =>
            getRepoId(org.organization, repo.name),
          ),
        )
        .flat(),
    )
    .reduce<string[]>((acc, repoName) => {
      if (acc.includes(repoName)) {
        throw new Error(`Duplicate repo: ${repoName}`)
      }
      return [...acc, repoName]
    }, [])
}

export class DefinitionFile {
  private path: string

  constructor(path: string) {
    this.path = path
  }

  public async getContents(): Promise<string> {
    return new Promise((resolve, reject) =>
      fs.readFile(this.path, "utf-8", (err, data) => {
        if (err) reject(err)
        else resolve(data)
      }),
    )
  }

  public async getDefinition(): Promise<Definition> {
    const definition = yaml.safeLoad(await this.getContents()) as Definition
    validateDefinition(definition)
    return definition
  }
}

export function getRepos(definition: Definition) {
  return definition.projects.flatMap((project) =>
    project.github
      .map((org) =>
        (org.repos || []).map((repo) => ({
          id: getRepoId(org.organization, repo.name),
          orgName: org.organization,
          project,
          repo,
        })),
      )
      .flat(),
  )
}

export function getGitHubOrgs(definition: Definition) {
  return uniq(
    definition.projects.flatMap((project) =>
      project.github.map((it) => it.organization),
    ),
  )
}
