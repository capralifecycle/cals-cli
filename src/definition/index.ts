export {
  DefinitionFile,
  getGitHubOrgs,
  getRepoId,
  getRepos,
} from "./definition"
export {
  Definition,
  DefinitionRepo,
  GetReposResponse,
  Project,
  RepoTeam,
} from "./types"
