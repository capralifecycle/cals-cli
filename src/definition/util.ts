import { Config } from "../config"

export function getDefinitionFile(config: Config) {
  if (process.env.CALS_DEFINITION_FILE) {
    return process.env.CALS_DEFINITION_FILE
  }

  return config.requireConfig("definitionFile")
}
