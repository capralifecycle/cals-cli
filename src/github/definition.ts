import fs from 'fs'
import yaml from 'js-yaml'
import { GitHubService } from './service'
import { Definition } from './types'

export function getDefinition(github: GitHubService) {
  const definition = yaml.safeLoad(getRawDefinition(github)) as Definition
  validateDefinition(definition)
  return definition
}

export function getRawDefinition(github: GitHubService) {
  return fs.readFileSync(github.getDefinitionFile(), 'utf-8')
}

function validateDefinition(definition: Definition) {
  // Verify no duplicates in users and extract known logins.
  const loginList = definition.users.reduce<string[]>((acc, user) => {
    if (acc.includes(user.login)) {
      throw new Error(`Duplicate login: ${user.login}`)
    }
    return [...acc, user.login]
  }, [])

  // Verify no duplicates in teams and extract team names.
  const teamNameList = definition.teams.reduce<string[]>((acc, team) => {
    if (acc.includes(team.name)) {
      throw new Error(`Duplicate team: ${team.name}`)
    }
    return [...acc, team.name]
  }, [])

  // Verify team members exists as users.
  definition.teams.forEach(team => {
    team.members.forEach(login => {
      if (!loginList.includes(login)) {
        throw new Error(
          `Team member ${login} in team ${
            team.name
          } is not registered in user list`,
        )
      }
    })
  })

  // Verify no duplicates in project names.
  definition.projects.reduce<string[]>((acc, project) => {
    if (acc.includes(project.name)) {
      throw new Error(`Duplicate project: ${project.name}`)
    }
    return [...acc, project.name]
  }, [])

  // Verify project teams exists as teams.
  definition.projects.forEach(project => {
    ;(project.teams || []).forEach(team => {
      if (!teamNameList.includes(team.name)) {
        throw new Error(
          `Project team ${team.name} in project ${
            project.name
          } is not registered in team list`,
        )
      }
    })
  })

  // Verify no duplicates in repos.
  definition.projects
    .flatMap(it => it.repos)
    .reduce<string[]>((acc, repo) => {
      if (acc.includes(repo.name)) {
        throw new Error(`Duplicate repo: ${repo.name}`)
      }
      return [...acc, repo.name]
    }, [])
}
