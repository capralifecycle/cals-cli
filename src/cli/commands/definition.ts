import {
  OrgsGetResponse,
  ReposGetResponse,
  ReposListTeamsResponseItem,
  TeamsListMembersResponseItem,
  TeamsListResponseItem,
} from '@octokit/rest'
import fs from 'fs'
import yaml from 'js-yaml'
import pAll from 'p-all'
import pMap from 'p-map'
import yargs, { CommandModule } from 'yargs'
import { Config } from '../../config'
import {
  getDefinition,
  getRawDefinition,
  getRepoId,
  getRepos,
} from '../../definition/definition'
import {
  Definition,
  DefinitionRepo,
  Project,
  RepoTeam,
  Team,
  User,
} from '../../definition/types'
import { createGitHubService, GitHubService } from '../../github/service'
import { Permission, Repo } from '../../github/types'
import { createSnykService, SnykService } from '../../snyk/service'
import { SnykGitHubRepo } from '../../snyk/types'
import { getGitHubRepo } from '../../snyk/util'
import { Reporter } from '../reporter'
import { createCacheProvider, createConfig, createReporter } from '../util'

interface DetailedProject {
  name: string
  repos: {
    [org: string]: {
      basic: Repo
      repository: ReposGetResponse
      teams: ReposListTeamsResponseItem[]
    }[]
  }
}

async function getReposFromGitHub(
  github: GitHubService,
  orgs: OrgsGetResponse[],
): Promise<DetailedProject['repos'][0]> {
  const queries = []

  for (const org of orgs) {
    const repos = await github.getRepoList({ owner: org.login })
    for (const repo of repos) {
      queries.push(async () => {
        const detailedRepo = await github.getRepository(
          repo.owner.login,
          repo.name,
        )
        if (detailedRepo === undefined) {
          throw Error(`Repo not found: ${repo.owner.login}/${repo.name}`)
        }

        return {
          basic: repo,
          repository: detailedRepo,
          teams: await github.getRepositoryTeamsList(detailedRepo),
        }
      })
    }
  }

  return await pAll(queries, { concurrency: 10 })
}

async function getTeams(github: GitHubService, orgs: OrgsGetResponse[]) {
  const result: {
    [org: string]: {
      team: TeamsListResponseItem
      members: TeamsListMembersResponseItem[]
    }[]
  } = {}

  for (const org of orgs) {
    const queries = []
    const teams = await github.getTeamList(org)
    for (const team of teams) {
      queries.push(async () => {
        return {
          team,
          members: await github.getTeamMemberList(team),
        }
      })
    }
    result[org.login] = await pAll(queries, { concurrency: 5 })
  }

  return result
}

function getCommonTeams(ownerRepos: DetailedProject['repos'][0]) {
  return ownerRepos.length === 0
    ? []
    : ownerRepos[0].teams.filter(team =>
        ownerRepos.every(repo =>
          repo.teams.some(
            otherTeam =>
              otherTeam.name === team.name &&
              otherTeam.permission === team.permission,
          ),
        ),
      )
}

function getSpecificTeams(
  teams: ReposListTeamsResponseItem[],
  commonTeams: ReposListTeamsResponseItem[],
) {
  return teams.filter(team => !commonTeams.some(it => it.name === team.name))
}

function getFormattedTeams(teams: ReposListTeamsResponseItem[]) {
  return teams.length === 0
    ? undefined
    : teams
        .map<RepoTeam>(it => ({
          name: it.name,
          permission: it.permission as Permission,
        }))
        .sort((a, b) => a.name.localeCompare(b.name))
}

async function getOrgs(github: GitHubService, orgs: string[]) {
  return pMap(orgs, it => github.getOrg(it), { concurrency: 5 })
}

function removeDuplicates<T, R>(items: T[], selector: (item: T) => R): T[] {
  const ids: R[] = []
  const result: T[] = []
  for (const item of items) {
    const id = selector(item)
    if (!ids.includes(id)) {
      result.push(item)
      ids.push(id)
    }
  }
  return result
}

async function getMembers(github: GitHubService, orgs: OrgsGetResponse[]) {
  return removeDuplicates(
    (await pMap(orgs, it => github.getOrgMembersList(it.login), {
      concurrency: 5,
    })).flat(),
    it => it.id,
  )
}

function buildTeamsList(
  list: Record<
    string,
    {
      team: TeamsListResponseItem
      members: TeamsListMembersResponseItem[]
    }[]
  >,
) {
  const result: Record<string, Team[]> = {}
  for (const [org, teams] of Object.entries(list)) {
    result[org] = teams
      // TODO: Only exclude if not referenced?
      .filter(it => it.members.length > 0)
      .map<Team>(team => ({
        name: team.team.name,
        members: team.members
          .map(it => it.login)
          .sort((a, b) => a.localeCompare(b)),
      }))
  }
  return result
}

async function dumpSetup(
  config: Config,
  reporter: Reporter,
  github: GitHubService,
  snyk: SnykService,
  outfile: string,
) {
  reporter.info('Fetching data. This might take some time')
  const orgs = await getOrgs(github, ['capralifecycle', 'capraconsulting'])
  const definition = getDefinition(config)
  const snykRepos = (await snyk.getProjects())
    .map(it => getGitHubRepo(it))
    .filter((it): it is SnykGitHubRepo => it !== undefined)
    .map(it => getRepoId(it.owner, it.name))
  const projectMap = getRepos(definition).reduce<Record<string, string>>(
    (acc, cur) => ({
      ...acc,
      [cur.id]: cur.project.name,
    }),
    {},
  )
  const repos = await getReposFromGitHub(github, orgs)

  const projects = Object.values(
    repos.reduce<{
      [project: string]: {
        name: string
        repos: {
          [owner: string]: typeof repos | undefined
        }
      }
    }>((acc, cur) => {
      const org = cur.repository.owner.login
      const projectName =
        projectMap[getRepoId(org, cur.repository.name)] || 'Unknown'
      const project = acc[projectName] || {
        name: projectName,
        repos: [],
      }

      return {
        ...acc,
        [projectName]: {
          ...project,
          repos: {
            ...project.repos,
            [org]: [...(project.repos[org] || []), cur],
          },
        },
      }
    }, {}),
  )
    .map<Project>(project => ({
      name: project.name,
      github: Object.entries(project.repos).reduce<Project['github']>(
        (acc, [org, list]) => {
          const commonTeams = getCommonTeams(list!)
          acc[org] = {
            teams: getFormattedTeams(commonTeams),
            repos: list!
              .map<DefinitionRepo>(repo => ({
                name: repo.basic.name,
                archived: repo.repository.archived ? true : undefined,
                issues: repo.repository.has_issues ? undefined : false,
                wiki: repo.repository.has_wiki ? undefined : false,
                teams: getFormattedTeams(
                  getSpecificTeams(repo.teams, commonTeams),
                ),
                snyk: snykRepos.includes(
                  getRepoId(repo.basic.owner.login, repo.basic.name),
                )
                  ? true
                  : undefined,
              }))
              .sort((a, b) => a.name.localeCompare(b.name)),
          }
          return acc
        },
        {},
      ),
    }))
    .sort((a, b) => a.name.localeCompare(b.name))

  const teams = await getTeams(github, orgs)
  const members = await getMembers(github, orgs)

  const generatedDefinition: Definition = {
    snyk: definition.snyk,
    github: {
      users: members
        .map<User>(
          it =>
            definition.github.users.find(user => user.login === it.login) || {
              type: 'external',
              login: it.login,
              // TODO: Fetch name from GitHub?
              name: '*Unknown*',
            },
        )
        .sort((a, b) => a.login.localeCompare(b.login)),
      teams: buildTeamsList(teams),
    },
    projects,
  }

  // TODO: An earlier version we had preserved comments by using yawn-yaml
  //  package. However it often produced invalid yaml, so we have removed
  //  it. We might want to revisit it to preserve comments.

  const doc = yaml.safeLoad(getRawDefinition(config))
  doc.snyk = generatedDefinition.snyk
  doc.projects = generatedDefinition.projects
  doc.github = generatedDefinition.github

  // Convert to/from plain JSON so that undefined elements are removed.
  fs.writeFileSync(outfile, yaml.safeDump(JSON.parse(JSON.stringify(doc))))
  reporter.info(`Saved to ${outfile}`)
}

const dumpSetupCommand: CommandModule = {
  command: 'dump-setup',
  describe:
    'Dump active setup as YAML. Will be formated same as the definition file.',
  builder: yargs =>
    yargs
      .positional('outfile', {
        type: 'string',
      })
      .demandOption('outfile'),
  handler: async argv => {
    const reporter = createReporter()
    const config = createConfig()
    const github = await createGitHubService(
      config,
      createCacheProvider(config),
    )
    const snyk = await createSnykService(config)
    await dumpSetup(config, reporter, github, snyk, argv.outfile as string)
  },
}

const command: CommandModule = {
  command: 'definition',
  describe: 'CALS definition file management',
  builder: yargs =>
    yargs.command(dumpSetupCommand).demandCommand().usage(`cals definition

The definition file is located at
https://github.com/capralifecycle/resources-definition/blob/master/resources.yaml

The file ~/.cals-config.json must include a reference to the location of this
file. For example by having this content:

  {
    "definitionFile": "/home/henrste/projects/capralifecycle/resources-definition/resources.yaml"
  }

Also remember to fetch the resources-definition repository every time you use cals-cli.`),
  handler: () => {
    yargs.showHelp()
  },
}

export default command
