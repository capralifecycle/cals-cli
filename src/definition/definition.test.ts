import del from "del"
import fs from "fs"
import tempy from "tempy"
import { DefinitionFile, schema } from "./definition"
import { Definition } from "./types"

describe("definition", () => {
  it("should error on reading invalid file", async () => {
    const tmp = tempy.file()
    await fs.promises.writeFile(
      tmp,
      JSON.stringify({
        some: "invalid",
      }),
    )

    const definitionFile = new DefinitionFile(tmp)

    await expect(
      definitionFile.getDefinition(),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Definition file invalid: data should have required property 'github', data should have required property 'projects'"`,
    )

    await del(tmp, { force: true })
  })

  it("should successfully parse correct file", async () => {
    const data: Definition = {
      github: {
        teams: [],
        users: [],
      },
      projects: [
        {
          name: "myproject",
          github: [
            {
              organization: "someorg",
              repos: [
                {
                  name: "somerepo",
                },
              ],
            },
          ],
        },
      ],
    }

    const tmp = tempy.file()
    await fs.promises.writeFile(tmp, JSON.stringify(data))

    const definitionFile = new DefinitionFile(tmp)
    await expect(definitionFile.getDefinition()).resolves.toStrictEqual(data)

    await del(tmp, { force: true })
  })

  it("should match expected schema", () => {
    // Keeping a snapshot of the schema helps us catch
    // unexpected changes.
    expect(schema).toMatchSnapshot()
  })
})
