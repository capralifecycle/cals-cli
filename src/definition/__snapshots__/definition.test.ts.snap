// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definition should match expected schema 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "DefinitionRepo": {
      "properties": {
        "archived": {
          "type": "boolean",
        },
        "issues": {
          "type": "boolean",
        },
        "name": {
          "type": "string",
        },
        "previousNames": {
          "items": {
            "$ref": "#/definitions/DefinitionRepoPreviousName",
          },
          "type": "array",
        },
        "public": {
          "type": "boolean",
        },
        "responsible": {
          "description": "Some external-defined entity being responsible for the repository.

Will override the project-defined responsible.",
          "type": "string",
        },
        "snyk": {
          "type": "boolean",
        },
        "teams": {
          "items": {
            "$ref": "#/definitions/RepoTeam",
          },
          "type": "array",
        },
        "wiki": {
          "type": "boolean",
        },
      },
      "required": [
        "name",
      ],
      "type": "object",
    },
    "DefinitionRepoPreviousName": {
      "properties": {
        "name": {
          "type": "string",
        },
        "project": {
          "type": "string",
        },
      },
      "required": [
        "name",
        "project",
      ],
      "type": "object",
    },
    "Permission": {
      "enum": [
        "admin",
        "pull",
        "push",
      ],
      "type": "string",
    },
    "Project": {
      "properties": {
        "github": {
          "items": {
            "properties": {
              "organization": {
                "type": "string",
              },
              "repos": {
                "items": {
                  "$ref": "#/definitions/DefinitionRepo",
                },
                "type": "array",
              },
              "teams": {
                "items": {
                  "$ref": "#/definitions/RepoTeam",
                },
                "type": "array",
              },
            },
            "required": [
              "organization",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "name": {
          "type": "string",
        },
        "responsible": {
          "description": "Some external-defined entity being responsible for the project.",
          "type": "string",
        },
        "tags": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": [
        "github",
        "name",
      ],
      "type": "object",
    },
    "RepoTeam": {
      "properties": {
        "name": {
          "type": "string",
        },
        "permission": {
          "$ref": "#/definitions/Permission",
        },
      },
      "required": [
        "name",
        "permission",
      ],
      "type": "object",
    },
    "Team": {
      "properties": {
        "members": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "name": {
          "type": "string",
        },
      },
      "required": [
        "members",
        "name",
      ],
      "type": "object",
    },
    "User": {
      "anyOf": [
        {
          "$ref": "#/definitions/UserBot",
        },
        {
          "$ref": "#/definitions/UserEmployee",
        },
        {
          "$ref": "#/definitions/UserExternal",
        },
      ],
    },
    "UserBot": {
      "properties": {
        "login": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "enum": [
            "bot",
          ],
          "type": "string",
        },
      },
      "required": [
        "login",
        "name",
        "type",
      ],
      "type": "object",
    },
    "UserEmployee": {
      "properties": {
        "capraUsername": {
          "type": "string",
        },
        "login": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "enum": [
            "employee",
          ],
          "type": "string",
        },
      },
      "required": [
        "capraUsername",
        "login",
        "name",
        "type",
      ],
      "type": "object",
    },
    "UserExternal": {
      "properties": {
        "login": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "type": {
          "enum": [
            "external",
          ],
          "type": "string",
        },
      },
      "required": [
        "login",
        "name",
        "type",
      ],
      "type": "object",
    },
  },
  "properties": {
    "github": {
      "properties": {
        "teams": {
          "items": {
            "properties": {
              "organization": {
                "type": "string",
              },
              "teams": {
                "items": {
                  "$ref": "#/definitions/Team",
                },
                "type": "array",
              },
            },
            "required": [
              "organization",
              "teams",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "users": {
          "items": {
            "$ref": "#/definitions/User",
          },
          "type": "array",
        },
      },
      "required": [
        "teams",
        "users",
      ],
      "type": "object",
    },
    "projects": {
      "items": {
        "$ref": "#/definitions/Project",
      },
      "type": "array",
    },
    "snyk": {
      "properties": {
        "accountId": {
          "type": "string",
        },
      },
      "required": [
        "accountId",
      ],
      "type": "object",
    },
  },
  "required": [
    "github",
    "projects",
  ],
  "type": "object",
}
`;
